### PR TITLE
Preserve Digit Separators During Reduce Tokens Stage.

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
Fix for Issue #22627 , by adding additional checks for digit separators… in floating point literals and decimals.

**Customer scenario**

```vb.net
Module Module1

    Sub Main()
        Dim z As Integer = 1_000_000
        Dim y As Single = 1_000.0_11
        Dim x As Double = 1_000.0_1_2
        Dim w As Decimal = 1_23_456.78_9D
    End Sub

End Module
```
The digit separators within the literal for variables `y`,`x`,`w` will be removed when moving the caret to another line. 

**Workarounds, if any**
Currently no know work around, to keep the digit separators present.

**Risk**
`Tiny`  

**Performance impact**
`Tiny` a couple of addition conditions within a workspace provider `ReduceTokensCodeCleanupProvider` 
This "Fix" disables formatting of the digits contains a separator.  Could be updated in the future to better handle them.

The current implementation removes digits separators, this pull-request attempts to preserves them, which it does but at a cost of ;-
  * no rounding
  * no removal of trail zeroes
  * no removal of leading zeroes

Fixes for those issues are discussed in (#22781).

**Is this a regression from a previous update?**
Unknown.

**Root cause analysis:**
 Initial thoughts were this was caused by the formatter, but I could not see where would be possible via visual inspection. So expanded the causal search to other sources.

### CODE REVIEW

File: src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests_WithDigitSeparators.cs [hunk 1]
```diff
@@ -0,0 +1,2137 @@
...
+        Const d_30_3 As Decimal = 1.234_567_890_123_456_789_012_345_678_91D         ' Dev11 & Roslyn: 1.2345678901234567890123456789D
+        Const d_30_4 As Decimal = 123_456_789_012_345.678_901_234_567_891D          ' Dev11 & Roslyn: 123456789012345.67890123456789D
+        Const d_30_5 As Decimal = 12_345_678_901_234_567_890_123_456_789.1D         ' Dev11 & Roslyn: 12345678901234567890123456789D
+
+        ' Overflow case 30 significant digits before decimal place: Ensure no pretty listing.
+        Const d_30_6 As Decimal = 123_456_789_012_345_678_901_234_567_891.0D        ' Dev11 & Roslyn: 123456789012345678901234567891.0D
+
+        Console.WriteLine(d_30_1)
+        Console.WriteLine(d_30_2)
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests_WithDigitSeparators.cs]: why are you running case correction and formatting?

Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests_WithDigitSeparators.cs]: we already have TestWorkspace for use in tests.  Is there a reason not to use that here?

File: src/Workspaces/VisualBasic/Portable/CodeCleanup/Providers/ReduceTokensCodeCleanupProvider.vb [hunk 2]
```diff
@@ -69,15 +74,13 @@ Namespace Microsoft.CodeAnalysis.CodeCleanup.Providers
                         End If
 
                         ' Compiler has parsed the literal text as a decimal value, fetch the string representation of this value.
-                        Dim valueText As String = GetDecimalLiteralValueString(value) + GetTypeCharString(literal.GetTypeCharacter())
+                        Dim valueText As String = GetDecimalLiteralValueString(value) & GetTypeCharString(literal.GetTypeCharacter())
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/CodeCleanup/Providers/ReduceTokensCodeCleanupProvider.vb]: why was this changed?

**Comment by CyrusNajmabadi** [comment]
tests are not actually passing on CI/CD.

File: src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests_WithDigitSeparators.cs [hunk 3]
```diff
@@ -0,0 +1,2137 @@
...
+        ' 7 significant digits
+        Const f_7_1 As Single = .149_999_5F      ' Dev11 & Roslyn: Pretty listed to 0.1499995F
+        Const f_7_2 As Single = 0.149_999_5f     ' Dev11 & Roslyn: Unchanged
+        Const f_7_3 As Single = 1.499_995F       ' Dev11 & Roslyn: Unchanged
+        Const f_7_4 As Single = 1_499.995f       ' Dev11 & Roslyn: Unchanged
+        Const f_7_5 As Single = 149_999.5F       ' Dev11 & Roslyn: Unchanged
+        Const f_7_6 As Single = 1_499_995.0f     ' Dev11 & Roslyn: Unchanged
+
+        Console.WriteLine(f_5_1)
```
[the remark this answers is not in the document]
Comment (AdamSpeight2008) [reply src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests_WithDigitSeparators.cs]: @CyrusNajmabadi I've used the existing [unit tests](
https://github.com/dotnet/roslyn/blob/master/src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests.cs#L46). Only altering them to use underscores.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/CoreTest/CodeCleanup/ReduceTokenTests_WithDigitSeparators.cs]: Please don't.  You made a trivial chnage.  We don't need to duplicate many thousands of lines of code just to validate that change.  A small handful of tests will suffice.  

Just to be explicit about this: when we make tiny code changes, we don't wholesale copy entire existing test suites, and then update all those tests to validate against those tiny changes.  We write the critical tests that validate that change in any particularly interesting ways.

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi The test succeed with CodeRush's Test Runner, not via the Test cmd.
`System.InvalidOperationException : Cannot create host services after test tear down.`

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi The test succeed with CodeRush's Test Runner, not via the Test cmd.

Can you update things so they work properly with Roslyn's CI system?  Thanks!
